consensus/polybft/state_sync_manager.go

```diff
@@ -0,0 +1,469 @@
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/umbracle/ethgo"
+	"google.golang.org/protobuf/proto"
+)
+
+// stateSyncConfig holds the configuration data of state sync manager
+type stateSyncConfig struct {
+	stateSenderAddr types.Address
+	jsonrpcAddr     string
+	dataDir         string
+	topic           topic
+	key             *wallet.Key
+}
+
+// StateSyncManager is a struct that manages the workflow of
+// saving and querying state sync events, and creating, and submitting new commitments
+type StateSyncManager struct {
+	// configuration fields
+	logger hclog.Logger
+	state  *State
+
+	config *stateSyncConfig
+
+	// per epoch fields
+	lock         sync.Mutex
+	commitment   *Commitment
+	epoch        uint64
+	validatorSet *validatorSet
+}
+
+// topic is an interface for a gossip p2p message
+type topic interface {
+	Publish(obj proto.Message) error
+	Subscribe(handler func(obj interface{}, from peer.ID)) error
+}
+
+// NewStateSyncManager creates a new instance of state sync manager
+func NewStateSyncManager(logger hclog.Logger, state *State, config *stateSyncConfig) (*StateSyncManager, error) {
+	s := &StateSyncManager{
+		logger: logger.Named("state-sync"),
+		state:  state,
+		config: config,
+		lock:   sync.Mutex{},
+	}
+
+	return s, nil
+}
+
+// init subscribes to bridge topics (getting votes) and start the event tracker routine
+func (s *StateSyncManager) init() error {
+	if err := s.initTracker(); err != nil {
+		return err
+	}
+
+	if err := s.initTransport(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// initTracker starts a new event tracker (to receive new state sync events)
+func (s *StateSyncManager) initTracker() error {
+	tracker := tracker.NewEventTracker(
+		path.Join(s.config.dataDir, "/deposit.db"),
+		s.config.jsonrpcAddr,
+		ethgo.Address(s.config.stateSenderAddr),
+		s,
+		s.logger)
+
+	return tracker.Start()
+}
+
+// initTransport subscribes to bridge topics (getting votes for commitments)
+func (s *StateSyncManager) initTransport() error {
+	return s.config.topic.Subscribe(func(obj interface{}, _ peer.ID) {
+		msg, ok := obj.(*polybftProto.TransportMessage)
+		if !ok {
+			s.logger.Warn("failed to deliver vote, invalid msg", "obj", obj)
+
+			return
+		}
+
+		var transportMsg *TransportMessage
+
+		if err := json.Unmarshal(msg.Data, &transportMsg); err != nil {
+			s.logger.Warn("failed to deliver vote", "error", err)
+
+			return
+		}
+
+		if err := s.saveVote(transportMsg); err != nil {
+			s.logger.Warn("failed to deliver vote", "error", err)
+		}
+	})
+}
+
+// saveVote saves the gotten vote to boltDb for later quorum check and signature aggregation
+func (s *StateSyncManager) saveVote(msg *TransportMessage) error {
+	s.lock.Lock()
+	epoch := s.epoch
+	valSet := s.validatorSet
+	s.lock.Unlock()
+
+	if valSet == nil {
+		// Epoch metadata is undefined
+		return nil
+	}
+
+	if msg.EpochNumber < epoch {
+		// or received message for some of the older epochs.
+		return nil
+	}
```

Comment:
Condition doesn't add much complexity written this way, but we spare a few lines of code :wink: 

```suggestion
	if valSet == nil || msg.EpochNumber < epoch {
		// Epoch metadata is undefined or received message for some of the older epochs.
		return nil
	}
```

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/1067/commits/acaaa3d20726fc29d0361e1b8472d936ceba4e67